File: libs/evals/src/gap/evals/runner/gap.py
"""GAP flow runner — stateless dispatch, envelope-based edits."""

from __future__ import annotations

import time
from pathlib import Path

from pydantic_ai import Agent
from pydantic_ai.models import Model

from ..agents import (
    StreamingLatency,
    clean_artifact,
    collect_structured_streaming_latency,
    collect_text_streaming_latency,
)
from ..apply import apply_envelope
from ..models import GAPTurnResult
from ..schema import LLMEnvelope


async def run_gap_turn0(
    llm: Model,
    init_system: str,
    turn0_prompt: str,
    output_dir: Path,
    ext: str,
) -> tuple[str, dict]:
    """Run turn-0 for the GAP flow (with target markers).

    Returns (artifact_text, metrics_dict).
    """
    agent: Agent[None, str] = Agent(llm, system_prompt=init_system)
    t0 = time.perf_counter()
    async with agent.run_stream(turn0_prompt) as r:
        raw_text, latency = await collect_text_streaming_latency(r)
        ms = int((time.perf_counter() - t0) * 1000)
        usage = r.usage()
        artifact = clean_artifact(raw_text)
        (output_dir / f"turn-0{ext}").write_text(artifact)

        metrics = {
            "input_tokens": usage.input_tokens,
            "output_tokens": usage.output_tokens,
            "latency_ms": ms,
            "artifact_bytes": len(artifact.encode()),
            "ttft_ms": latency.ttft_ms,
            "ttlt_ms": latency.ttlt_ms,
            "median_itl_ms": latency.median_itl_ms,
        }
        return artifact, metrics


async def run_gap_flow(
    llm: Model,
    maintain_system: str,
    artifact: str,
    edit_prompts: list[tuple[str, str]],
    fmt: str,
    output_dir: Path,
    ext: str,
) -> tuple[list[GAPTurnResult], str]:
    """Run all edit turns for the GAP flow.

    Returns (per_turn_results, final_artifact).
    """
    maintain_agent: Agent[None, LLMEnvelope] = Agent(
        llm,
        system_prompt=maintain_system,
        output_type=LLMEnvelope,
    )

    results: list[GAPTurnResult] = []
    version = 1

    for turn_name, edit_prompt in edit_prompts:
        turn_num = int(turn_name.split("-")[1])
        user_msg = (
            f"## Current Artifact\n\n```\n{artifact}\n```\n\n"
            f"## Edit Instruction\n\n{edit_prompt}"
        )

        t0 = time.perf_counter()
        parsed = False
        succeeded = False
        env_name = ""
        envelope_json = ""
        latency = StreamingLatency()
        usage = type("U", (), {"input_tokens": 0, "output_tokens": 0})()

        try:
            async with maintain_agent.run_stream(user_msg) as r:
                envelope, latency = await collect_structured_streaming_latency(r)
                ms = int((time.perf_counter() - t0) * 1000)
                usage = r.usage()

                parsed = True
                env_name = envelope.name
                envelope_json = envelope.model_dump_json(indent=2)

                new_artifact = apply_envelope(artifact, envelope, fmt)
                succeeded = True
                artifact = new_artifact
                version += 1

        except Exception as e:
            ms = int((time.perf_counter() - t0) * 1000)

        if envelope_json:
            (output_dir / f"{turn_name}.json").write_text(envelope_json)
        (output_dir / f"{turn_name}{ext}").write_text(artifact)

        results.append(GAPTurnResult(
            turn=turn_num,
            edit=edit_prompt[:80],
            input_tokens=usage.input_tokens,
            output_tokens=usage.output_tokens,
            latency_ms=ms,
            output_bytes=len(artifact.encode()),
            ttft_ms=latency.ttft_ms,
            ttlt_ms=latency.ttlt_ms,
            median_itl_ms=latency.median_itl_ms,
            envelope_parsed=parsed,
            apply_succeeded=succeeded,
            envelope_name=env_name,
            failed=not succeeded,
            failure_reason="" if succeeded else "parse or apply failed",
        ))

    return results, artifact

File: apps/eval-cli/runner/gap.go
package runner

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"

	agent "github.com/urmzd/saige/agent"
	agenteval "github.com/urmzd/saige/agent/eval"
	"github.com/urmzd/saige/agent/types"

	"github.com/urmzd/generative-artifact-protocol/eval-cli/apply"
	"github.com/urmzd/generative-artifact-protocol/eval-cli/gap"
)

// RunGAPTurn0 runs turn-0 for the GAP flow (with target markers in the system prompt).
func RunGAPTurn0(ctx context.Context, provider types.Provider, initSystem, turn0Prompt, outputDir, ext string) (string, Turn0Metrics, error) {
	ag := agent.NewAgent(agent.AgentConfig{
		SystemPrompt: initSystem,
		Provider:     provider,
		MaxIter:      1,
	})

	t0 := time.Now()
	stream := ag.Invoke(ctx, []types.Message{types.NewUserMessage(turn0Prompt)})
	timing, rawText, deltas := agenteval.CollectStreamTiming(stream.Deltas())
	if err := stream.Wait(); err != nil {
		return "", Turn0Metrics{}, fmt.Errorf("gap turn-0: %w", err)
	}
	if streamErr := extractStreamError(deltas); streamErr != nil {
		return "", Turn0Metrics{}, fmt.Errorf("gap turn-0: %w", streamErr)
	}
	ms := int(time.Since(t0).Milliseconds())

	artifact := gap.CleanArtifact(rawText)
	_ = os.WriteFile(filepath.Join(outputDir, "turn-0"+ext), []byte(artifact), 0o644)

	metrics := Turn0Metrics{
		InputTokens:   timing.InputTokens,
		OutputTokens:  timing.OutputTokens,
		LatencyMs:     ms,
		ArtifactBytes: len(artifact),
	}
	if timing.TTFTMs > 0 {
		metrics.TTFTMs = &timing.TTFTMs
	}
	if timing.TTLTMs > 0 {
		metrics.TTLTMs = &timing.TTLTMs
	}
	if timing.MedianITL > 0 {
		metrics.MedianITLMs = &timing.MedianITL
	}

	return artifact, metrics, nil
}

// RunGAPFlow runs all edit turns for the GAP flow.
// Each turn is stateless: the LLM receives only the current artifact + edit instruction
// and returns a structured LLMEnvelope, which is applied via the Rust engine.
func RunGAPFlow(ctx context.Context, provider types.Provider, maintainSystem, artifact string, editPrompts []gap.TurnPrompt, format, outputDir, ext string) ([]gap.GAPTurnResult, string, error) {
	schema := types.SchemaFrom[gap.LLMEnvelopeSchema]()

	var results []gap.GAPTurnResult
	version := 1

	for _, tp := range editPrompts {
		turnNum := parseTurnNum(tp.Name)
		userMsg := fmt.Sprintf("## Current Artifact\n\n```\n%s\n```\n\n## Edit Instruction\n\n%s", artifact, tp.Prompt)

		// Fresh agent per turn (stateless).
		ag := agent.NewAgent(agent.AgentConfig{
			SystemPrompt:   maintainSystem,
			Provider:       provider,
			MaxIter:        1,
			ResponseSchema: &schema,
		})

		t0 := time.Now()
		stream := ag.Invoke(ctx, []types.Message{types.NewUserMessage(userMsg)})
		timing, rawText, deltas := agenteval.CollectStreamTiming(stream.Deltas())
		err := stream.Wait()
		if err == nil {
			err = extractStreamError(deltas)
		}
		ms := int(time.Since(t0).Milliseconds())

		parsed := false
		succeeded := false
		envName := ""
		envelopeJSON := ""

		if err == nil {
			// Parse the structured output as an Envelope.
			var envelope gap.Envelope
			if jsonErr := json.Unmarshal([]byte(rawText), &envelope); jsonErr == nil {
				parsed = true
				envName = envelope.Name
				prettyJSON, _ := json.MarshalIndent(envelope, "", "  ")
				envelopeJSON = string(prettyJSON)

				// Apply the envelope to the artifact.
				newArtifact, applyErr := apply.ApplyEnvelope(artifact, []byte(rawText), format, envelope.ID, envelope.Version)
				if applyErr == nil {
					succeeded = true
					artifact = newArtifact
					version++
				}
			}
		}

		if envelopeJSON != "" {
			_ = os.WriteFile(filepath.Join(outputDir, tp.Name+".json"), []byte(envelopeJSON), 0o644)
		}
		_ = os.WriteFile(filepath.Join(outputDir, tp.Name+ext), []byte(artifact), 0o644)

		tr := gap.TurnResult{
			Turn:         turnNum,
			Edit:         truncate(tp.Prompt, 80),
			InputTokens:  timing.InputTokens,
			OutputTokens: timing.OutputTokens,
			LatencyMs:    ms,
			OutputBytes:  len(artifact),
			Failed:       !succeeded,
		}
		if !succeeded {
			tr.FailureReason = "parse or apply failed"
		}
		if timing.TTFTMs > 0 {
			tr.TTFTMs = &timing.TTFTMs
		}
		if timing.TTLTMs > 0 {
			tr.TTLTMs = &timing.TTLTMs
		}
		if timing.MedianITL > 0 {
			tr.MedianITLMs = &timing.MedianITL
		}

		results = append(results, gap.GAPTurnResult{
			TurnResult:     tr,
			EnvelopeParsed: parsed,
			ApplySucceeded: succeeded,
			EnvelopeName:   envName,
		})
	}

	return results, artifact, nil
}
